blog/_src/posts/2017-11-16-monotonicity-types-towards-a-type-system-for-eventual-consistency.md

```diff
@@ -0,0 +1,97 @@
+    Title: Monotonicity Types: Towards A Type System for Eventual Consistency
+    Date: 2017-10-22T11:59:06
+    Tags: types, monotonicity, CRDTs, eventual consistency, by Kevin Clancy
+
+A few weeks back, we published a draft of an article entitled [_Monotonicity Types_](https://infoscience.epfl.ch/record/231867). In it, we describe a type system which we hope can aid the design of distributed systems by tracking monotonicity with types.
+
+But first, what, precisely, do we mean by _monotonicity_? Here's a short definition:
+
+A partially ordered set is a set \\(P\\) endowed with a relation \\(\leq\\) such that for all \\(p, q, r \in P\\) we have:
+
+1. \\(p \leq p\\) (reflexivity)
+2. \\(p \leq  q\\) and \\(q \leq r\\) implies \\(p \leq r\\) (transitivity)
+3. \\(p \leq q\\) and \\(q \leq p\\) implies \\(p = q\\) (anti-symmetry)
+
+If \\(P\\) and \\(Q\\) are partially ordered sets, we say that a function \\(f : P \to Q\\)
+between them is *monotone* if for all \\(p_1, p_2 \in P\\) with \\(p_1 \leq p_2\\), we have \\(f(p_1) \leq f(p_2)\\).
+
+So, said another way, increasing the input to a monotone function causes an increase to its output.
```

Comment:
Well not really ... the function could always return the top element. How about changing to:

> increasing the input to a monotone function increases the lower bound on its output

Reply:
A couple of remarks. First, I think Kevin's description is better for building intuition, while Ben's one is somewhat “heavy”. Second, there is quite a tradition (dates back to Bourbaki, afaik) to use “increasing” for sequences (functions) which are really not-decreasing. In contrast to the notion of “strictly increasing”. This view is justified by [Wikipedia](https://en.wikipedia.org/wiki/Sequence#Increasing_and_decreasing), if I'm not mistaken. (And, well, by Bourbaki: Set theory, Ch. 1, n. 5).